### PR TITLE
Suggest calling `Self::associated_function()`

### LINKED ISSUE #96453: suggest calling Self::associated_function()
Given the following code: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=72d0a7426f82aeca7615c923eb4d2da4

```rust
struct S;

impl S {
    fn helper() {}

    fn doer(&self) {
        helper();
    }
}
```

The current output is:

```
Compiling playground v0.0.1 (/playground)
error[[E0425]](https://doc.rust-lang.org/stable/error-index.html#E0425): cannot find function `helper` in this scope
 [--> src/lib.rs:7:9
](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=72d0a7426f82aeca7615c923eb4d2da4#)  |
7 |         helper();
  |         ^^^^^^ not found in this scope

For more information about this error, try `rustc --explain E0425`.
error: could not compile `playground` due to previous error
```

Ideally the output should look like:

```
Consider calling the associated function `Self::helper()`.
```

### PR DESCRIPTION
**Pull request by TaKO8Ki**
closes #96453

### CODE REVIEW

File: compiler/rustc_resolve/src/late/diagnostics.rs [hunk 1]
```diff
@@ -171,46 +173,83 @@ impl<'a: 'ast, 'ast> LateResolutionVisitor<'a, '_, 'ast> {
...
+                );
+                debug!(
+                    "self.diagnostic_metadata.current_function={:?}",
+                    self.diagnostic_metadata.current_function
+                );
+                let suggestion = if let Some(items) = self.diagnostic_metadata.current_impl_items
+                    && let Some((fn_kind, _)) = self.diagnostic_metadata.current_function
+                    && self.current_trait_ref.is_none()
+                    && let Some(function_ctxt) = fn_kind.ctxt()
+                    && let FnCtxt::Assoc(_) = function_ctxt
```
Comment (lcnr) [inline compiler/rustc_resolve/src/late/diagnostics.rs:193]:
```suggestion
                    && let Some(FnCtxt::Assoc(_)) = fn_kind.ctxt()
```
does that work?

File: compiler/rustc_resolve/src/late/diagnostics.rs [hunk 2]
```diff
@@ -171,46 +173,83 @@ impl<'a: 'ast, 'ast> LateResolutionVisitor<'a, '_, 'ast> {
...
+                    && let Some(item) = items.iter().find(|i| {
+                        if let AssocItemKind::Fn(fn_) = &i.kind
+                            && fn_.sig.decl.inputs.is_empty()
```
Comment (lcnr) [inline compiler/rustc_resolve/src/late/diagnostics.rs:196]: why do you only suggest that if there are no inputs?

I am not too familiar with this function

File: compiler/rustc_resolve/src/late/diagnostics.rs [hunk 3]
```diff
@@ -147,10 +147,12 @@ impl<'a: 'ast, 'ast> LateResolutionVisitor<'a, '_, 'ast> {
         let mut expected = source.descr_expected();
         let path_str = Segment::names_to_string(path);
         let item_str = path.last().unwrap().ident;
-        let (base_msg, fallback_label, base_span, could_be_expr) = if let Some(res) = res {
+        let (base_msg, fallback_label, base_span, could_be_expr, suggestion) = if let Some(res) =
```
Comment (lcnr) [inline compiler/rustc_resolve/src/late/diagnostics.rs:150]: i dislike large tuples like this. it's just personal preference, but what's your take on moving these 4 fields to a helper struct?

**Review by lcnr — COMMENTED** [review]
some small questions

File: compiler/rustc_resolve/src/late/diagnostics.rs (hunk 3 above)
Comment (TaKO8Ki) [reply compiler/rustc_resolve/src/late/diagnostics.rs:150]: I agree with you. I'll fix it just like you said.

File: compiler/rustc_resolve/src/late/diagnostics.rs (hunk 2 above)
Comment (TaKO8Ki) [reply compiler/rustc_resolve/src/late/diagnostics.rs]: I fixed it to suggest if inputs don't include `self`, `&self`, or the `&mut self`.

File: compiler/rustc_resolve/src/late/diagnostics.rs [hunk 4]
```diff
@@ -171,45 +178,91 @@ impl<'a: 'ast, 'ast> LateResolutionVisitor<'a, '_, 'ast> {
...
-                (String::new(), "this scope".to_string())
+            let (mod_prefix, mod_str, suggestion) = if path.len() == 1 {
+                debug!(
+                    "self.diagnostic_metadata.current_impl_items={:?}",
+                    self.diagnostic_metadata.current_impl_items
+                );
+                debug!(
+                    "self.diagnostic_metadata.current_function={:?}",
+                    self.diagnostic_metadata.current_function
+                );
```
Comment (lcnr) [inline compiler/rustc_resolve/src/late/diagnostics.rs:193]:
```suggestion
                debug!(?self.diagnostic_metadata.current_impl_items);
                debug!(?self.diagnostic_metadata.current_function);
```

File: compiler/rustc_resolve/src/late/diagnostics.rs [hunk 5]
```diff
@@ -171,45 +178,91 @@ impl<'a: 'ast, 'ast> LateResolutionVisitor<'a, '_, 'ast> {
...
+                    && let Some(item) = items.iter().find(|i| {
+                        if let AssocItemKind::Fn(fn_) = &i.kind
+                            && fn_.sig.decl.inputs.iter().all(|i| match &i.ty.kind {
+                                TyKind::Rptr(_, ty) => match ty.ty.kind {
+                                    TyKind::ImplicitSelf => false,
+                                    _ => true,
+                                }
+                                TyKind::ImplicitSelf => false,
+                                _ => true
+                            })
```
Comment (lcnr) [inline compiler/rustc_resolve/src/late/diagnostics.rs:207]:
```suggestion
                            && !fn_.sig.decl.has_self()
```

File: compiler/rustc_resolve/src/late/diagnostics.rs [hunk 6]
```diff
@@ -171,45 +178,91 @@ impl<'a: 'ast, 'ast> LateResolutionVisitor<'a, '_, 'ast> {
...
+                            && i.ident.name == item_str.name
+                        {
+                            debug!("name={}", item_str.name);
+                            debug!("fn_.sig.decl.inputs={:?}", fn_.sig.decl.inputs);
```
Comment (lcnr) [inline compiler/rustc_resolve/src/late/diagnostics.rs:211]:
```suggestion
                            debug!(?item_str.name);
                            debug!(?fn_.sig.decl.inputs);
```
or completely remove these debug statements '^^

**Review by lcnr — COMMENTED** [review]
after that r=me

**Comment by lcnr** [comment]
Thanks :+1: 

@bors r+ rollup

**Pull request merged**